agents/plastic_agent.py

```diff
@@ -0,0 +1,86 @@
+import pandas as pd
+import numpy as np
+from sklearn import preprocessing
+class PlasticAgent():
+    '''
+    create a Plastic Policy Agent
+    -Input required- Models provided in csv format, with two columns- state and action taken by agent in that state, each row is a sample from which we extract transition function
+    -param models: List of models provided
+    -param num_models: Number of models for creating Behavior Distribution
+    -param tf_probs: Extract transition function from list of models
+    -param BehaviorDist: Probability Distribution over all models
+    -param eta: Value of Eta, default=0.2
+    '''
+    def __init__(self, num_models, models):
+        self.models = models
+        self.num_models = num_models
+        self.tf_probs = self.extract_trans()
+        self.BehaviorDist = self.init_BehaviorDist(self.num_models)
+        print("initial distribution:", self.BehaviorDist)
+        self.eta = 0.2
+    
+    def extract_trans(self):
+        # Initialize empty transition function for each model
+        tf_probs = []
+        for model in self.models:
+            # Initialize empty transition function for this model
+            tf = pd.DataFrame(columns = ['state', 'action', '#instances'])
+            # iterate over all model rows
+            for index, row in model.iterrows():
+                flag = 0
+                if(len(tf.index)>0):
+                    for id,row_tf in tf.iterrows():
+                        # If this state and action already existed, add to instances count
+                        if row.at['state'] == row_tf.at['state'] and row.at['p1_curr_subtask'] == row_tf.at['action']:
+                            row_tf.at['#instances'] = row_tf.at['#instances'] + 1
+                            flag = 1
+                            break
+                if flag !=1:
+                    tf.loc[len(tf.index)] = [row.at['state'], row.at['p1_curr_subtask'], 1]
+            tf_prob = pd.DataFrame(columns = ['state', 'action', '#instances'])
+            for index,row in tf.iterrows():
+                # for each state, sum instances over all actions and calculate probability of taking each action from that state
+                state_tf = tf.loc[tf['state'] == row.at['state']]
+                tf = tf[tf['state'] != row.at['state']]
+                total_instances = state_tf[['#instances']].sum()
+                state_tf['#instances'] = state_tf['#instances'].apply(lambda x: x/total_instances.iloc[0])
+                tf_prob = tf_prob.append(state_tf)
+            tf_probs.append(tf_prob)
+        return tf_probs
+    
+    def init_BehaviorDist(self,n):
+        # initialize initial Probability Distribution as uniform
+        return [1/n]*n 
+
+    def UpdateBelief(self,state,action):
+        for index, model in enumerate(self.models) :
+            # Initialize P(action|model,state)
+            prob_from_tf = 0
```

Comment:
with the dict setup, this can be simplified to:
```
action_idx = Action.ACTION_TO_INDEX[action]
prob_from_tf = model[state][action_idx] if state in model else 0
```
However, the real way we want to use this is using ml models, so what we eventually want is:
 ```
action_idx = Action.ACTION_TO_INDEX[action]
obs = self.encoding_fn(self.env.mdp, state, self.grid_shape, self.args.horizon, p_idx=p_idx)
prob_dist = model.get_distribution(obs)
action_prob = prob_dist[action_idx]
 ```
 
 For the time being, maybe make a flag to switch between the static distributions (what you are currently using) and ml models